contracts/sidechain/PausableProxyOFT.sol

```diff
@@ -0,0 +1,193 @@
+       Core 
+    ------------------------------------------------------------------- */
+
+    /**
+     * @dev Override sendFrom to add pause modifier
+     */
+    function sendFrom(
+        address _from,
+        uint16 _dstChainId,
+        bytes calldata _toAddress,
+        uint256 _amount,
+        address payable _refundAddress,
+        address _zroPaymentAddress,
+        bytes calldata _adapterParams
+    ) public payable override whenNotPaused {
+        outflow[_getCurrentEpoch()] += _amount;
+        super.sendFrom(_from, _dstChainId, _toAddress, _amount, _refundAddress, _zroPaymentAddress, _adapterParams);
+    }
+
+    /**
+     * @dev Override _sendAck on OFTCore
+     *
+     * Add functionality to
+     * 1) Pause the bridge
+     * 2) Add an inflow limit safety net. This safety net should never get hit in
+     *    normal market operations. It's intended to mitigate risk in case of a
+     *    doomsday event.
+     */
+    function _sendAck(
+        uint16 _srcChainId,
+        bytes memory _srcAddress,
+        uint64 _nonce,
+        bytes memory _payload
+    ) internal override {
+        (, bytes memory toAddressBytes, uint256 amount) = abi.decode(_payload, (uint16, bytes, uint256));
+
+        uint256 epoch = _getCurrentEpoch();
+        uint256 currInflow = inflow[epoch];
+        uint256 newInflow = currInflow.add(amount);
+        inflow[epoch] = newInflow;
+        address to = toAddressBytes.toAddress(0);
+
+        if (_getNetInflow(newInflow, outflow[epoch]) > inflowLimit || paused()) {
+            // If the net inflow is greater than the limit for this epoch OR the bridge
+            // is currently paused we send each transfer to a queue for delayed processing.
+            // In the case of a doomsday event this limits the exposure to the bridge.
+            queue[keccak256(abi.encode(epoch, _srcChainId, to, amount, block.timestamp))] = true;
+            emit QueuedFromChain(epoch, _srcChainId, to, amount, block.timestamp);
+        } else {
+            // Process the transfer as normal
+            amount = _creditTo(_srcChainId, to, amount);
+            emit ReceiveFromChain(_srcChainId, to, amount);
+        }
+    }
+
+    /**
+     * @dev Process a queued transfer
+     *      Transfer has to be a valid root and the queue delay has to have passed
+     * @param _epoch        Epoch
+     * @param _srcChainId   Source chain ID
+     * @param _to           Address to transfer to
+     * @param _amount       Amount to transfer
+     * @param _timestamp    Time when this transfer was queued
+     */
+    function processQueued(
+        uint256 _epoch,
+        uint16 _srcChainId,
+        address _to,
+        uint256 _amount,
+        uint256 _timestamp
+    ) external whenNotPaused {
+        bytes32 queueRoot = keccak256(abi.encode(_epoch, _srcChainId, _to, _amount, _timestamp));
+        require(queue[queueRoot], "!root");
```

Comment:
crit : if we don't dequeue processed hashes, we've got an unlimited mint bug waiting to happen!